### PR TITLE
Code analyzer to detect if route parameters are parsable

### PR DESCRIPTION
**Pull request by mitchdenny**
# Code analyzer to detect if route parameters are parsable

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Code Analyzer to detect whether route parameters are parsable (have a TryParse method or implement IParsable<TSelf>).

## Description

This PR adds to the ```RouteHandlerAnalyzer``` by analyzing the route pattern passed in on endpoint ```MapXXX``` methods and then checking that the types that they map to are parsable (either by implementing ```IParsable<TSelf>``` or just exposing a ```TryParse``` as is the convention.

This PR only partially addresses the issue linked below because there is some ambiguity on how we handle the ```BindAsync``` scenario when parameters on the handler may either be bindable from ```HttpContext``` OR injectable via a service. Figured it was better to get the parsable scenario tucked away first which does not have this ambiguity.

Partially addresses #35733

## TODOs

- ~~[ ] Flag when developer sets [FromBody] on parameter AND type implements ```BindAsync``` (to discuss)~~

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=45286) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=45286) before we can accept your contribution.<br/><br/>:x: mitchdenny [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=45286)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=45286) it.</sub>

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:72]: You don't need to do this for some built-in types. Test `SpecialType` for boolean and string.

See https://github.com/dotnet/aspnetcore/blob/4ff55988a7f634aa443693bdd917488b9b018cd9/src/Framework/AspNetCoreAnalyzers/src/Analyzers/Http/HeaderDictionaryIndexerAnalyzer.cs#L30

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:41]: I believe the logic for IsParsable already exists in another analyzer. Check out:

https://github.com/dotnet/aspnetcore/blob/4ff55988a7f634aa443693bdd917488b9b018cd9/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs#L380-L455

I think this logic can be moved to a common location and shared between analyzers.

Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs]: Lifted this code into a Common namespace and transplanted a few of the types there. Take a look at let me know if that aligns with what you were thinking. I also extracted some of the methods that were already present around determining parsability and augmented them a little to handle the ```IParsable<TSelf>``` scenario.

File: src/Framework/App.Ref/src/CompatibilitySuppressions.xml
Comment (danmoseley) [reply src/Framework/App.Ref/src/CompatibilitySuppressions.xml:1]: What decides to write them to disk - shouldn't it be a human deciding to suppress something?

Comment (dougbu) [reply src/Framework/App.Ref/src/CompatibilitySuppressions.xml:1]: Also, why are we creating compatibility suppression files in CI builds❔ Suggest suppressing the file writes except in local builds.

> What decides to write them to disk - shouldn't it be a human deciding to suppress something?

The `$(GenerateCompatibilitySuppressionFile)` property but I don't see the point given the suppressions are **not** in our code and don't seem to add much value.

More generally, humans at least _can_ decide what goes into their PRs. `git add .` and `git commit --all` allow people to take shortcuts however.

Comment (wtgodbe) [reply src/Framework/App.Ref/src/CompatibilitySuppressions.xml:1]: We could just flip `GenerateCompatibilitySuppressionFile` to off. That's been there since the initial checkin of us using the Package Validation SDK

**Comment by mitchdenny** [comment]
@JamesNK I've swapped over to using ```RoutePatternParser``` instead of ```RouteTokenEnumerator```. I'm loathed to go any further with this PR until your WellKnownTypes refactor has gone in because my next step is reusing some of the logic around special parameter types on handler delegates.

Either I merge first and you feel some pain or I wait until you merge to avoid feeling even more pain ;)

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs:18]:
```suggestion
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs:162]:
```suggestion
        foreach (var methodSymbol in bindAsyncMethods)
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs:116]:
```suggestion
    }

    private static bool IsBindableViaIBindableFromHttpContext(ITypeSymbol typeSymbol)
```

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:49]: Accepting this change but doing manually in my workspace due to a merge issue.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:13]: Done in local workspace.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RouteWellKnownTypes.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RouteWellKnownTypes.cs:34]: No, we don't in fact. I don't know why I did this. Removing.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypes.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypes.cs:79]: Removed most of removing special case handling for numerics/temporals. Retained ```ParameterInfo``` and ```ValueTask<T>``` though.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs:165]: Fixed in local workspace.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs]: Added a method although it doesn't simplify it too much due to the number of parameters being passed around ;)

File: src/Framework/App.Ref/src/CompatibilitySuppressions.xml
Comment (mitchdenny) [reply src/Framework/App.Ref/src/CompatibilitySuppressions.xml]: Will deal with this in a separate PR.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:336]: This should work. The `source` being tested with a POCO `Customer` is perfectly valid.

```csharp
webApp.MapPost(
    "/customers",
    ([FromBody] Customer customer) => {
    });
```

`[FromBody] Customer customer` and just `Customer customer` have the same behavior in this case assuming there's no `Customer` registered as a service. It's deserialized as JSON. Uploading `{}` as the body works just fine.

Even if `Customer` did implement `BindAsync`, `[FromBody]` is enough to force JSON deserialization (unless the parameter type is `Stream` or `PipeReader`). We check for the `[FromBody]` attribute here. We could rename `BindParameterFromBody` to `BindParameterFromJsonBody` in this PR to clarify that code a bit.

https://github.com/dotnet/aspnetcore/blob/600eb9aa53c052ec7327e2399744215dbe493a89/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L699-L713

And only fall back to BindAsync later here:

https://github.com/dotnet/aspnetcore/blob/600eb9aa53c052ec7327e2399744215dbe493a89/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L798-L801

I'm just linking to the RDF code for reference. I wouldn't expect anyone to figure this out just from reading the code.  However, it would be good to run these sample apps to verify it really is broken. It's also probably be a good idea to update the docs at https://github.com/dotnet/AspNetCore.Docs/blob/96d01d63c1886691c96fafa4a299f01948a8ff34/aspnetcore/fundamentals/minimal-apis/includes/parameter-binding.md#binding-precedence to clarify that `[FromBody]` means JSON deserialization, Stream or PipeReader and not a parameter type with a `BindAsync` method which may or may not read from the body.

That said, and this isn't the case for this test method, it might make sense to use `BindAsync` if it exists even when there is a `[FromBody]` attribute. We don't do that today, but I bet people would expect us to, and I doubt anyone is relying on us *not* calling `BindAsync` in that case.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:361]: This is the case I just mentioned above where this will work, but not the way you want. It will try to deserilize as JSON because of the `[FromBody]` attribute.

We should raise a diagnostic if there is no `[FromBody]` attribute because this returns a `Task` instead of `ValueTask`. It took me a second to figure out why this was supposed to be broken. I wonder if we can do even better than saying `"Parameter '{0}' of type {1} should define a ValueTask<{1}> BindAsync(HttpContext) or implement IBindableFromHttpContext<{1}>."` in this case. I can imagine a lot of people thinking "that's what I did!"

Maybe `"BindAsync must return a ValueTask<{1}> instead of Task<{1}>. Consider implementing IBindableFromHttpContext<{1}>."` for this specific case?

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (halter73) [inline src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:397]: We probably should have an analyzer flag this case because we're going to try to deserialize `Customer` as JSON because of the `[FromBody]` attribute and that's probably not what people want if they're implementing `BindAsync`.

**Comment by halter73** [comment]
Also, what's the behavior when `IParsable<>` and/or `IBindableFromHttpContext<>` are implemented explicitly? Will this raise diagnostics about `TryParse` and/or `BindAsync` being incorrectly implmented? It probably should because we won't find them at runtime today, but it shouldn't as soon as #43546 gets fixed.

**Comment by mitchdenny** [comment]
> Also, what's the behavior when `IParsable<>` and/or `IBindableFromHttpContext<>` are implemented explicitly?

Right now, it won't raise a diagnostic because once I detect that either of the interfaces are present, I don't go any further. Is it better to ship this analyzer before or after the fix for #43546 is in?

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:75]: These shouldn't check for the attribute directly. What matters is `FromHeaderAttribute` implements `IFromHeaderMetadata`.

Check out where `Microsoft_AspNetCore_Http_Metadata_IFromHeaderMetadata` and friends is used.

**Comment by JamesNK** [comment]
> Is it better to ship this analyzer before or after the fix for #43546 is in?

Before is fine. That issue will get fixed sometime for .NET 8.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:333]: > `[FromBody] Customer customer` and just `Customer customer` have the same behavior in this case assuming there's no `Customer` registered as a service. It's deserialized as JSON. Uploading `{}` as the body works just fine.

Note quite true from my experimentation:

```(Customer customer) => {}``` will use ```BindAsync(...)``` if it is available and fallback to JSON deserialize if it isn't.
```([FromBody]Customer customer) => {}``` will always JSON deserialize.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:394]: I wonder how likely it is that folks might use the same type on different routes differently (JSON serialization works one way and they rely on ```BindAsync``` for another behavior). I guess they should implement a separate DTO in that situation.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:358]: Will need another diagnostic but I agree with this. When I was first building out these test cases I got it wrong ;)

**Comment by halter73** [comment]
> Is it better to ship this analyzer before or after the fix for #43546 is in?

Don't hold up the analyzer on the runtime issue. Worst case there's some false negatives, but we should fix the runtime behavior in 8 regardless.

However, I would like tests with explicit interface implementations, so we don't accidentally change behavior.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (halter73) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:394]: That's technically a possibility, and that flexibility one way our current behavior could be seen as "better". I don't think it's super realistic to want to support both JSON deserialization and custom `BindAsync` type, so I think this flexibility probably causes more confusion than it's worth.

I wouldn't patch the behavior because it's been that way since .NET 6, but I'd definitely consider changing it for 8.

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (halter73) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:333]: You're correct. I was focusing on the case where there is no `BindAsync` or `TryParse` implementation like in this test, but I didn't make that clear.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs:99]: Is there a reason for comparing via name instead of `SymbolEqualityComparer` here?

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/ParsabilityHelper.cs:109]: name vs SymbolEqualityComparer

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs:377]:
```suggestion
    private static bool IsCurrentParameterBindable(SyntaxToken token, SemanticModel semanticModel, WellKnownTypes wellKnownTypes, CancellationToken cancellationToken)
```
Was this changed to internal at some point during dev? Only accept if it can be changed back.

**Review by JamesNK — APPROVED** [review]

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:360]: At the moment I've got it nope'ing out of anything for the ```[FromBody]``` scenario since the current behavior is that it will either work via JSON deserialization OR via BindAsync (if present). I think the throwing up an error/warning when BindAsync is not required, but does exist but doesn't conform to the expected interface is a bridge too far since its not like BindAsync is a reserved method name or anything - thoughts?

Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:360]: > Also, what's the behavior when `IParsable<>` and/or `IBindableFromHttpContext<>` are implemented explicitly? Will this raise diagnostics about `TryParse` and/or `BindAsync` being incorrectly implmented? It probably should because we won't find them at runtime today, but it shouldn't as soon as #43546 gets fixed.

Do you think this is worth doing given that implementing interfaces explicitly is not that common and we actually want to change this behavior in the databinding plumbing anyway? I think this analyzer is going to ship with 8.0, I'm assuming that we'll have this explicit interface support in for 8.0?

File: src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs
Comment (mitchdenny) [reply src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs:397]: OK, I've dropped in a more specific diagnostic for this scenario. Took a little while to plumb it together but it means if we want to start getting really helpful with these diagnostics we can.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:24]: Using the indexer like this to get an argument value doesn't work with named parameters. e.g.
```cs
endpoints.MapGet(delegate: () => {}, pattern: "/helloworld");
```

I suggest this instead:
```cs
    private static bool TryGetStringToken(IInvocationOperation invocation, out SyntaxToken token)
    {
        IArgumentOperation? argumentOperation = null;
        foreach (var argument in invocation.Arguments)
        {
            if (argument.Parameter?.Ordinal == 1)
            {
                argumentOperation = argument;
            }
        }

        if (argumentOperation?.Syntax is not ArgumentSyntax routePatternArgumentSyntax ||
            routePatternArgumentSyntax.Expression is not LiteralExpressionSyntax routePatternArgumentLiteralSyntax)
        {
            token = default;
            return false;
        }

        token = routePatternArgumentLiteralSyntax.Token;
        return true;
    }
```

**Comment by JamesNK** [comment]
Please squash merge. We use it in aspnetcore to keep the commit history cleaner.

**Comment by mitchdenny** [comment]
Apologies, I had assumed we used policy to enforce that (I never even think about it when pressing merge as a result). My bad.

**Pull request merged**
